File: 백준/Silver/11725. 트리의 부모 찾기/트리의 부모 찾기.cc
#include <bits/stdc++.h>

#define endl '\n'
#define FASTIO ios::sync_with_stdio(0), cin.tie(0), cout.tie(0)

void _init() {
#ifndef BOJ
      freopen("boj.in", "r", stdin);
#endif // BOJ
}

using namespace std;

using vi = vector<int>;

int n;
vector<vi> adj;
vi p;

void bfs() {
    queue<int> q;
    q.push(1);
    while (!q.empty()) {
        //cout << q.size() << endl;
        int cur = q.front();
        q.pop();
        for (int nxt : adj[cur]) {
            if (nxt == p[cur]) continue;
            p[nxt] = cur;
            q.push(nxt);
        }
    }
}

int main() {
    FASTIO;
    _init();

    cin >> n;
    adj = vector<vi>(n + 1);
    p = vi(n + 1);
    for (int i = 0; i < n - 1; i++) {
        int u, v; cin >> u >> v;
        adj[u].push_back(v);
        adj[v].push_back(u);
    }

    bfs();
    
    for (int i = 2; i <= n; i++) cout << p[i] << endl;

    return 0;
}

File: 백준/Silver/11725. 트리의 부모 찾기/트리의 부모 찾기.py
import sys
# FASTIO -> input은 꽤 느린 편이기 때문에 sys.stdin.readline 사용, 자세한건 GPT한테
input = lambda: sys.stdin.readline().rstrip()
sys.setrecursionlimit(int(2e5)) # 재귀 깊이 제한 넉넉하게 설정


def dfs(cur, par):
    """
    cur: 현재 방문하는 노드
    par: cur의 부모 노드
    """
    for nxt in adj[cur]:
        if nxt == par: continue
        parent[nxt] = cur
        dfs(nxt, cur)
        

n = int(input())
adj = [[] for _ in range(n+1)]
for _ in range(n-1):
    u, v = map(int, input().split())
    adj[u].append(v)
    adj[v].append(u)
parent = [-1] * (n+1)
dfs(1, -1)
for i in range(2, n+1):
    print(parent[i])
